airbyte-integrations/connectors/source-quickbooks/source_quickbooks/components.py

```diff
@@ -0,0 +1,63 @@
+import datetime
+import typing
+from collections import abc
+from dataclasses import dataclass
+
+import dpath.util
+from airbyte_cdk.sources.declarative.incremental import DatetimeBasedCursor
+from airbyte_cdk.sources.declarative.types import StreamSlice, Record
+
+
+class LastRecordDictProxy(abc.MutableMapping):
+    """
+    Patch a dict object to be able to get/set/delete/etc... values by path.
+    Example:
+        >>> record = LastRecordDictProxy({"root": {"nested": "value"}})
+        >>> record["root/nested"]
+        <<< "value"
+        >>> record.get("root/nested")
+        <<< "value"
+    """
+
+    def __init__(self, record: Record, field_mapping: typing.Mapping = None):
+        self._record: Record = record
+        self._field_mapping = field_mapping if field_mapping is not None else {}
+
+    def __setitem__(self, k: str, v: typing.Any) -> None:
+        dpath.util.set(self._record, k, v)
+
+    def __delitem__(self, v: str) -> None:
+        dpath.util.delete(self._record, v)
+
+    def __getitem__(self, k: str) -> typing.Any:
+        try:
+            return dpath.util.get(self._record, k)
+        except KeyError as e:
+            if k in self._field_mapping:
+                return dpath.util.get(self._record, self._field_mapping[k])
+            raise e
+
+    def __len__(self) -> int:
+        return len(self._record)
+
+    def __iter__(self) -> typing.Iterator:
+        return self._record
+
+    def __bool__(self):
+        return bool(self._record)
+
+
+@dataclass
+class CustomDatetimeBasedCursor(DatetimeBasedCursor):
```

Comment:
please follow the process outlined [here](https://docs.google.com/document/d/1LB4gBk0slHj3zNmq1Qtj1U7VkKyTN9eFsVldgQ8T4HQ/edit) for recording issues for these components

Reply:
done